form/src/main/kotlin/com/ritense/form/service/FormSubmissionService.kt

```diff
@@ -0,0 +1,314 @@
+        } catch (notFoundException: DocumentNotFoundException) {
+            logger.error("Document could not be found", notFoundException)
+            FormSubmissionResultFailed(FromException(notFoundException))
+        } catch (notFoundException: ProcessDocumentDefinitionNotFoundException) {
+            logger.error("ProcessDocumentDefinition could not be found", notFoundException)
+            FormSubmissionResultFailed(FromException(notFoundException))
+        } catch (ex: RuntimeException) {
+            val referenceId = UUID.randomUUID()
+            logger.error("Unexpected error occurred - {}", referenceId, ex)
+            FormSubmissionResultFailed(
+                OperationError.FromString("Unexpected error occurred, please contact support - referenceId: $referenceId")
+            )
+        }
+    }
+
+    private fun getProcessDocumentDefinition(
+        processLink: ProcessLink,
+        document: Document?
+    ): ProcessDocumentDefinition {
+        val processDefinition = processService.getProcessDefinitionById(processLink.processDefinitionId)
+        val processDefinitionKey = CamundaProcessDefinitionKey(processDefinition.key)
+        return if (document == null) {
+            processDocumentAssociationService.getProcessDocumentDefinition(processDefinitionKey)
+        } else {
+            processDocumentAssociationService.getProcessDocumentDefinition(
+                processDefinitionKey,
+                document.definitionId().version()
+            )
+        }
+    }
+
+    private fun getProcessVariables(taskInstanceId: String?): JsonNode? {
+        return if (!taskInstanceId.isNullOrEmpty()) {
+            Mapper.INSTANCE.get().valueToTree(camundaTaskService.getTaskVariables(taskInstanceId))
+        } else {
+            null
+        }
+    }
+
+    private fun getFormFields(
+        formDefinition: FormIoFormDefinition,
+        formData: JsonNode
+    ): List<FormField> {
+        return formDefinition.documentMappedFieldsForSubmission
+            .mapNotNull { objectNode -> FormField.getFormField(formData, objectNode, applicationEventPublisher) }
+    }
+
+    private fun getSubmittedDocumentContent(formFields: List<FormField>, document: Document?): ObjectNode {
+        val submittedDocumentContent = JsonNodeFactory.instance.objectNode()
+        formFields.forEach {
+            it.preProcess(document)
+            it.appendValueToDocument(submittedDocumentContent)
+        }
+        return submittedDocumentContent
+    }
+
+    private fun getPreJsonPatch(
+        formDefinition: FormIoFormDefinition,
+        submittedDocumentContent: JsonNode,
+        processVariables: JsonNode?,
+        document: Document?
+    ): JsonPatch {
+        //Note: Pre patch can be refactored into a specific field types that apply itself
+        val preJsonPatch = prefillFormService.preSubmissionTransform(
+            formDefinition,
+            submittedDocumentContent,
+            processVariables,
```

Comment:
Can't you create an empty JsonNode when the placeholders or content are null?

Reply:
changed